source/module_basis/module_nao/pswfc_radials.cpp

```diff
@@ -0,0 +1,308 @@
+    }
+
+    itype_ = itype;
+    read_upf_pswfc(ifs, screening_coeff, conv_thr, ptr_log, rank);
+    set_rcut_max();
+    
+    if (rank == 0)
+    {
+        ifs.close();
+    }
+}
+
+bool PswfcRadials::startswith(std::string word, std::string pattern)
+{
+    if(word.size() < pattern.size()) return false;
+    int score = 1;
+    for(int ic = 0; ic < pattern.size(); ic++)
+    {
+        if(word[ic] != pattern[ic])
+        {
+            score *= 0;
+        }
+        else
+        {
+            score *= 1;
+        }
+    }
+    return bool(score);
+}
+
+std::string PswfcRadials::steal_from_quotes(std::string word)
+{
+    // first make sure there are even number of quotes in this word
+    int num_quote = 0;
+    for(auto letter: word)
+    {
+        if(letter == '\"') num_quote += 1;
+    }
+    assert(num_quote % 2 == 0);
+    // then steal from quotes
+    std::string result;
+    size_t _left = word.find_first_of("\"");
+    size_t _right = word.find_last_of("\"");
+    result = word.substr(_left+1, _right-_left-1);
+    // then remove all spaces ahead
+    while(result[0] == ' ')
+    {
+        result.erase(0, 1);
+    }
+    return result;
+}
+
+std::string PswfcRadials::steal_from_quotes(std::ifstream& ifs, std::string word)
+{
+    // concatenate all words until the second quote, no matter how many lines and spaces between
+    std::string concatenated = word.substr(
+        word.find_first_of("\"")+1, 
+        word.size()-word.find_first_of("\"")-1
+        );
+    int num_quote = 1;
+    while(num_quote < 2)
+    {
+        std::string line;
+        ifs >> line;
+        for(auto letter: line)
+        {
+            if(letter == '\"') num_quote += 1;
+            if(num_quote == 2) break;
+            concatenated += letter;
+        }
+    }
+    // then remove all spaces ahead
+    while(concatenated[0] == ' ')
+    {
+        concatenated.erase(0, 1);
+    }
+    return concatenated;
+}
+
+std::string PswfcRadials::read_keyword_value(std::ifstream& ifs, std::string word)
+{
+    // count the number of quotes, only 1 or 2 cases are considered for pseudopotential reading
+    int num_quote = 0;
+    for(auto letter: word)
+    {
+        if(letter == '\"') num_quote += 1;
+    }
+    assert(num_quote == 1 || num_quote == 2);
+    if(num_quote == 1) return steal_from_quotes(ifs, word);
+    else return steal_from_quotes(word);
+}
+
+double PswfcRadials::radial_norm(const std::vector<double> rgrid,
+                                 const std::vector<double> rvalue)
+{
+    std::vector<double> integrand(rvalue.size());
+    for(int ir = 0; ir != rvalue.size(); ++ir)
+    {
+        integrand[ir] = rvalue[ir] * rvalue[ir] * rgrid[ir] * rgrid[ir];
+    }
+    double dr = rgrid[1] - rgrid[0];
+    double norm = ModuleBase::Integral::simpson(rvalue.size(), integrand.data(), dr);
+    norm = sqrt(norm);
+    return norm;
+}
+
+double PswfcRadials::cut_to_convergence(const std::vector<double>& rgrid,
```

Comment:
similar with hydrogen-like orbital that is generated segment-by-segment, because pswfc initialy defined in pseudopotential file speads in a quite wide range, the rcut of it is too large and will bring about high computational costs, results in mismatch between number of kpoints and supercells